### DISCUSSION TITLE
Conditional content based on format

### DISCUSSION THREAD

**Discussion by snhansen** [quarto-dev/quarto-cli#8205] [Q&A] [answered]
### Description

Consider the code below:

    ---
    format: pdf
    ---
    
    Something here
    
    ::: {.content-visible when-format="html"}
    
    ```{r}
    library(modelsummary)
    datasummary_skim(mtcars, type = "numeric", output = "html")
    ```
    
    :::

I'm getting an error rendering this because the `datasummary_skim()` call produces html output. But why is this an issue when the whole code block shouldn't be compiled in the final .pdf document because of the content-visible thing? What am I missing here?

**Comment by mcanouil** [comment]
It seems to be an issue with the R package output which triggers Quarto error, when you set `output = "html"`.
If you don't set this option, you get the proper results for each formats.

````qmd
---
format: pdf
---

Something here

::: {.content-visible when-format="html"}

```{r}
knitr::kable(head(mtcars), format = "html")
```

:::
````


FYI, the error is the following:

```sh
Error: Functions that produce HTML output found in document targeting pdf output.
Please change the output type of this document to HTML.
If your aiming to have some HTML widgets shown in non-HTML format as a screenshot,
please install webshot or webshot2 R package for knitr to do the screenshot.
Alternatively, you can allow HTML output in non-HTML formats
by adding this option to the YAML front-matter of
your quarto file:

  prefer-html: true

Note however that the HTML output will not be visible in non-HTML formats.
```

I am not exactly sure why the error is triggered here (https://github.com/quarto-dev/quarto-cli/blob/main/src/execute/rmd.ts).
Note that this is completely independent of conditional content div as the error can be triggered without it.
CC @cderv

**Reply by cderv** [reply, marked as answer]
This is expected error by **knitr** and **rmarkdown** because HTML code is produced in output while the output format is seen to be PDF. 

So this will be detected at computation rendering time (during knitting) and an error is thown. In some cases when **webshot** or **webshot2** is detected, a screenshot of the HTML rendered is taken (for leaflet for example). 

So basically this happens because 

* Quarto conditional content happens after computation. Everything in the document will be computed. 
* **knitr** and **rmarkdown** does throw an error when HTML code is produced in a non-HTML output like pdf. 
* It doesn't know about the code chunk being wrapped in  conditional quarto div

This will be improved when we'll adapt conditional to apply before computation rendering. 

Current knitr workaround would be the usual `#| eval: !expr knitr::is_html_output()` instead of `content-visible`

**Comment by snhansen** [comment]
Oh, that's weird. Thanks for the reply. Removing the option solves it, and will in fact also render correctly to .pdf.

**Reply by mcanouil** [reply]
FYI: the following is enough to trigger Quarto error

````qmd
---
format: pdf
---

```{r}
library(modelsummary)
datasummary_skim(mtcars, type = "numeric", output = "html")
```
````

**Reply by cderv** [reply]
It is not quarto, this is rmarkdown and knitr directly. `output = "html"` means to write the Table in HTML and the format output is PDF so expected LaTeX. 

HTML code in LaTeX output means nothing, so an error is thrown and this is expected. I'll comment futher in the other answer

**Reply by mcanouil** [reply]
Well, knitr::kable does not have the issue.
Also the message seems to come from https://github.com/quarto-dev/quarto-cli/blob/main/src/execute/rmd.ts

````qmd
---
format: pdf
---

Something here

::: {.content-visible when-format="html"}

```{r}
knitr::kable(head(mtcars), format = "html")
```

:::
````

**Reply by cderv** [reply]
`knitr::kable()` does not produce HTML dependencies. I did not go deep into the explanation, but if you go look in the **rmarkdown** code you'll find what triggers this. 

So to be clearer: the error is triggered for any code cell output that has some HTML dependencies to be included in the output. HTML in PDF output is not expected, so it triggers an error. 

I can assure you that the message comes from R Markdown 
https://github.com/rstudio/rmarkdown/blob/65916c33c36b126be06fce2e2f134965028dc266/R/render.R#L799-L812

**Reply by cderv** [reply]
> Removing the option solves it, and will in fact also render correctly to .pdf.

@snhansen by the way, when you don't set the option will use the default 
https://modelsummary.com/man/datasummary_skim.html#arguments

and it is possible that the default do the right thing and adapt its output to the output format. (i.e creating a LaTeX table in PDF and HTML table in html). Set `keep-md: true` to see what it will do in each case

**Reply by snhansen** [reply]
Yeah, that looks to be the case indeed. Didn't know it was that clever. Thanks for the help, both of you.

**Comment by cderv** [comment]
> But why is this an issue when the whole code block shouldn't be compiled in the final .pdf document because of the content-visible thing? What am I missing here?

To be clear on the why an error is thrown - Current limitation is that Conditional Content in Quarto using the Markdown syntax will apply at the end and computation of all cells with `eval: true` still happens. 

It happens that there is mechanism with knitr to detect HTML output from chunk in non-HTML format rendering. 

This is why you see the issue.
